chainer/functions/connection/bilinear.py

```diff
@@ -81,7 +81,12 @@ def forward(self, inputs):
 
         xp = cuda.get_array_module(*inputs)
         if xp is numpy:
-            y = numpy.einsum('ij,ik,jkl->il', e1, e2, W)
+            # optimize: y = numpy.einsum('ij,ik,jkl->il', e1, e2, W)
+            y = numpy.einsum(
+                'ij,ijl->il',
+                e1,
+                numpy.einsum('ik,jkl->ijl', e2, W)
```

Comment:
For instance, this line can be replaced with a single tensordot over axes 1 in both e2 and W.